uqcsbot/advent.py

```diff
@@ -183,296 +288,654 @@ def __init__(self, bot: UQCSBot):
-        """
-        Returns a string representing the global leaderboard of the given list.
+        if code == UQCS_LEADERBOARD:
+            message = ":star: *Advent of Code UQCS Leaderboard* :trophy:"
+        else:
+            message = f":star: *Advent of Code Leaderboard {code}* :trophy:"
 
-        Full leaderboard includes rank, global points, and username.
-        """
+        if day:
+            message += f"\n:calendar: *Day {day}* (Sorted By {sortby})"
+            members = [member for member in members if member.attempted_day(day)]
+            members.sort(key=lambda m: sorting_functions_for_day[sortby](m, day))
+        else:
+            members = [
+                member
+                for member in members
+                if any(member.attempted_day(day) for day in ADVENT_DAYS)
+            ]
+            members.sort(key=sorting_functions_for_month[sortby])
 
-        #   3     4
-        # |-|  |--|
-        #   1)  751 Name
-        def format_member(i: int, m: Member):
-            return f"{i:>3}) {m.global_:>4} {m.name}"
+        if not members:
+            await interaction.edit_original_response(
+                content="This leaderboard contains no people."
+            )
+            return
 
-        return "\n".join(format_member(i, m) for i, m in enumerate(members, 1))
+        scoreboard_file = io.BytesIO(
+            bytes(
+                print_leaderboard(
+                    parse_leaderboard_column_string(leaderboard_style, self.bot),
+                    members,
+                    day,
+                ),
+                "utf-8",
+            )
+        )
+        await interaction.edit_original_response(
+            content=message,
+            attachments=[
+                discord.File(
+                    scoreboard_file,
+                    filename=f"advent_{code}_{year}_{day}.txt",
+                )
+            ],
+        )
 
-    def format_day_leaderboard(self, members: List[Member]) -> str:
+    @advent_command_group.command(name="register")
+    @app_commands.describe(
+        aoc_name="Your name shown on Advent of Code.",
+    )
+    async def register_command(self, interaction: discord.Interaction, aoc_name: str):
         """
-        Returns a string representing the leaderboard of the given members on
-        the given day.
-
-        Full leaderboard includes rank, points, stars (per day), and username.
+        Register for prizes by linking your discord to an Advent of Code name.
         """
+        # TODO: Check UQCS membership
+        await interaction.response.defer(thinking=True)
 
-        def format_seconds(seconds: Optional[int]) -> str:
-            if seconds is None:
-                return ""
-            delta = timedelta(seconds=seconds)
-            if delta > timedelta(hours=24):
-                return ">24h"
-            return str(delta)
+        id = self._get_unused_registration_id()
```

Comment:
The aim was to allow for different registrations each year (say if you're using a different AOC name). I'm happy to remove for simplicity.

Also, will do with the DB;s generating methods. I was just copying `remindme.py` (it looked ugly, but I assumed that was standard practice).

Reply:
It'll need a funky migration because you have to manually `ALTER TABLE` to set it up, but it's worth it. I wasn't aware remindme also did this; that should probably be changed too.

Is it possible to use a different AoC name for different years? I guess if you switch to a different OAuth account and don't migrate your old progress it's a thing, but at that point, you've made the decision to unlink your old progress from your current account anyway.